src/chain/optimism.rs

```diff
@@ -0,0 +1,286 @@
+//! Optimism
+#![allow(missing_docs)]
+use std::collections::{BTreeMap, HashMap};
+
+use alloy_chains::NamedChain;
+use alloy_consensus::{Signed, TxEip1559, TxEip2930, TxEip4844, TxLegacy};
+use alloy_primitives::{Bytes, B256, U256};
+use alloy_rpc_types::{BlockTransactions, Header};
+use op_alloy_consensus::{OpDepositReceipt, OpReceiptEnvelope, OpTxEnvelope, OpTxType, TxDeposit};
+use op_alloy_network::eip2718::Encodable2718;
+use op_alloy_rpc_types::Transaction;
+use revm::{
+    primitives::{BlockEnv, OptimismFields, SpecId, TxEnv},
+    Handler,
+};
+
+use crate::{mv_memory::MvMemory, BuildIdentityHasher, MemoryLocation, PevmTxExecutionResult};
+
+use super::{PevmChain, RewardPolicy};
+
+/// Implementation of [PevmChain] for Optimism
+#[derive(Debug, Clone, PartialEq)]
+pub struct PevmOptimism {
+    id: u64,
+}
+
+impl PevmOptimism {
+    pub fn mainnet() -> Self {
+        PevmOptimism {
+            id: NamedChain::Optimism.into(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum OptimismBlockSpecError {
+    MissingBlockNumber,
+    UnsupportedSpec,
+}
+
+/// Represents errors that can occur when parsing transactions
+#[derive(Debug, Clone, PartialEq)]
+pub enum OptimismTransactionParsingError {
+    ConversionError(String),
+    InvalidType(u8),
+    MissingGasPrice,
+    MissingMaxFeePerGas,
+    MissingSourceHash,
+    OverflowedGasLimit,
+    SerdeError(String),
+    UnexpectedType(u8),
+}
+
+fn get_optimism_gas_price(tx: &Transaction) -> Result<U256, OptimismTransactionParsingError> {
+    let tx_type_raw = tx.inner.transaction_type.unwrap_or_default();
+    let Ok(tx_type) = OpTxType::try_from(tx_type_raw) else {
+        return Err(OptimismTransactionParsingError::InvalidType(tx_type_raw));
+    };
+
+    match tx_type {
+        OpTxType::Legacy | OpTxType::Eip2930 => tx
+            .inner
+            .gas_price
+            .map(U256::from)
+            .ok_or(OptimismTransactionParsingError::MissingGasPrice),
+        OpTxType::Eip1559 | OpTxType::Eip4844 => tx
+            .inner
+            .max_fee_per_gas
+            .map(U256::from)
+            .ok_or(OptimismTransactionParsingError::MissingMaxFeePerGas),
+        OpTxType::Deposit => Ok(U256::ZERO),
+    }
+}
+
+/// Convert [Transaction] to [OptimismFields]
+pub(crate) fn get_optimism_fields(
```

Comment:
These functions are non-trivial. Let's add a URL to OP's doc or an implementation in op-geth/op-reth/alloy/etc as reference.